File: lib/web/js/request.js
import * as ops from 'request.go'
import { Blob } from 'web:blob'

const METHODS = ['DELETE', 'GET', 'HEAD', 'OPTIONS', 'POST', 'PUT']
const kRequestId = Symbol('requestId')
const kHeaders = Symbol('headers')
const kBodyUsed = Symbol('bodyUsed')
const kHasBody = Symbol('hasBody')

const DEFAULTS = {
    method: 'GET',
    mode: 'cors',
    credentials: 'same-origin',
    cache: 'default',
    redirect: 'follow',
    referrer: 'about:client',
    integrity: '',
    keepalive: false,
}

const encoder = new TextEncoder()
const decoder = new TextDecoder()

function normalizeMethod(method = 'GET') {
    const token = String(method).toUpperCase()
    return METHODS.includes(token) ? token : token
}

function headersToRecord(headers) {
    const record = {}
    if (!headers) {
        return record
    }

    const assignPair = (key, value) => {
        const normalizedKey = String(key).toLowerCase()
        record[normalizedKey] = String(value)
    }

    if (typeof headers.forEach === 'function') {
        headers.forEach((value, key) => {
            assignPair(key, value)
        })
        return record
    }

    if (typeof headers.entries === 'function') {
        const entries = headers.entries()
        if (typeof entries?.[Symbol.iterator] === 'function') {
            for (const [key, value] of entries) {
                assignPair(key, value)
            }
            return record
        }

        if (Array.isArray(entries)) {
            for (const pair of entries) {
                if (Array.isArray(pair) && pair.length === 2) {
                    assignPair(pair[0], pair[1])
                }
            }
            return record
        }
    }

    if (Array.isArray(headers)) {
        for (const pair of headers) {
            if (Array.isArray(pair) && pair.length === 2) {
                assignPair(pair[0], pair[1])
            }
        }
        return record
    }

    if (typeof headers === 'object') {
        for (const key of Object.keys(headers)) {
            assignPair(key, headers[key])
        }
    }

    return record
}

function recordToHeaders(record) {
    return new Headers(record ?? {})
}

function normalizeURL(input) {
    if (typeof input === 'string') {
        return new URL(input).href
    }

    if (input instanceof URL) {
        return input.href
    }

    throw new TypeError('Request input must be a string or URL')
}

function hasOverrides(init) {
    if (!init) {
        return false
    }

    const overrideKeys = [
        'method',
        'headers',
        'body',
        'mode',
        'credentials',
        'cache',
        'redirect',
        'referrer',
        'integrity',
        'keepalive'
    ]

    return overrideKeys.some((key) => init[key] !== undefined)
}

function collectBlobPartIds(blob) {
    const ids = []

    const walk = (target) => {
        if (!(target instanceof Blob)) {
            return
        }

        const parts = target._parts ?? []
        for (const part of parts) {
            if (part instanceof Blob) {
                walk(part)
            } else if (part && typeof part._id === 'string') {
                ids.push(part._id)
            }
        }
    }

    walk(blob)
    return ids
}

function normalizeBody(body, headers) {
    if (body === null || body === undefined) {
        return { kind: 'none' }
    }

    if (body instanceof Blob) {
        if (body.type && !headers.has('content-type')) {
            headers.set('content-type', body.type)
        }
        const partIds = collectBlobPartIds(body)
        const descriptor = {
            kind: 'blob',
            blobParts: partIds,
            contentType: body.type || null,
        }
        return descriptor
    }

    if (body instanceof ArrayBuffer) {
        return { kind: 'bytes', bytes: new Uint8Array(body) }
    }

    if (ArrayBuffer.isView(body)) {
        return {
            kind: 'bytes',
            bytes: new Uint8Array(body.buffer, body.byteOffset, body.byteLength)
        }
    }

    if (typeof body === 'string') {
        if (!headers.has('content-type')) {
            headers.set('content-type', 'text/plain;charset=UTF-8')
        }
        return { kind: 'bytes', bytes: encoder.encode(body), contentType: headers.get('content-type') }
    }

    if (body instanceof URLSearchParams) {
        if (!headers.has('content-type')) {
            headers.set('content-type', 'application/x-www-form-urlencoded;charset=UTF-8')
        }
        return { kind: 'bytes', bytes: encoder.encode(body.toString()), contentType: headers.get('content-type') }
    }

    if (body instanceof Uint8Array) {
        return { kind: 'bytes', bytes: new Uint8Array(body) }
    }

    throw new TypeError('Unsupported Request body type')
}

function createSnapshot(metadata, bodyDescriptor, headers) {
    return {
        url: metadata.url,
        method: metadata.method,
        headers: headersToRecord(headers),
        credentials: metadata.credentials,
        mode: metadata.mode,
        cache: metadata.cache,
        redirect: metadata.redirect,
        referrer: metadata.referrer,
        integrity: metadata.integrity,
        keepalive: metadata.keepalive,
        bodyUsed: false,
        hasBody: bodyDescriptor.kind !== 'none',
        contentType: headers.get('content-type') ?? bodyDescriptor.contentType ?? '',
    }
}

function createRequestFromSnapshot(id, snapshot) {
    const instance = Object.create(Request.prototype)
    instance[kRequestId] = id
    instance._url = snapshot.url
    instance._method = snapshot.method
    instance[kHeaders] = recordToHeaders(snapshot.headers)
    instance._credentials = snapshot.credentials ?? DEFAULTS.credentials
    instance._mode = snapshot.mode ?? DEFAULTS.mode
    instance._cache = snapshot.cache ?? DEFAULTS.cache
    instance._redirect = snapshot.redirect ?? DEFAULTS.redirect
    instance._referrer = snapshot.referrer ?? DEFAULTS.referrer
    instance._integrity = snapshot.integrity ?? DEFAULTS.integrity
    instance._keepalive = Boolean(snapshot.keepalive)
    instance[kBodyUsed] = Boolean(snapshot.bodyUsed)
    instance[kHasBody] = Boolean(snapshot.hasBody)
    instance._contentType = snapshot.contentType ?? ''
    return instance
}

async function consumeBodyBytes(request) {
    if (request[kBodyUsed]) {
        throw new TypeError('Body has already been used for this Request')
    }

    request[kBodyUsed] = true
    const result = ops.op_request_consume_body(request[kRequestId])

    if (!result) {
        return new Uint8Array()
    }

    if (result instanceof Uint8Array) {
        return result
    }

    if (result instanceof ArrayBuffer) {
        return new Uint8Array(result)
    }

    if (ArrayBuffer.isView(result)) {
        return new Uint8Array(result.buffer, result.byteOffset, result.byteLength)
    }

    return Uint8Array.from(result)
}

class Request {
    constructor(input, init = {}) {
        if (input instanceof Request && !hasOverrides(init)) {
            if (input.bodyUsed) {
                throw new TypeError('Cannot construct Request with disturbed body')
            }
            const clonedId = ops.op_request_clone(input[kRequestId])
            const snapshot = ops.op_request_snapshot(clonedId)
            return createRequestFromSnapshot(clonedId, snapshot)
        }

        if (input instanceof Request && hasOverrides(init)) {
            throw new TypeError('Constructing Request from Request with overrides is not supported yet')
        }

        const url = normalizeURL(input)
        const method = normalizeMethod(init.method ?? DEFAULTS.method)

        const headers = new Headers(init.headers ?? {})
        const bodyDescriptor = normalizeBody(init.body ?? null, headers)

        if ((method === 'GET' || method === 'HEAD') && bodyDescriptor.kind !== 'none') {
            throw new TypeError('Body not allowed for GET or HEAD requests')
        }

        const metadata = {
            url,
            method,
            headers: headersToRecord(headers),
            credentials: init.credentials ?? DEFAULTS.credentials,
            mode: init.mode ?? DEFAULTS.mode,
            cache: init.cache ?? DEFAULTS.cache,
            redirect: init.redirect ?? DEFAULTS.redirect,
            referrer: init.referrer ?? DEFAULTS.referrer,
            integrity: init.integrity ?? DEFAULTS.integrity,
            keepalive: Boolean(init.keepalive ?? DEFAULTS.keepalive),
            bodyKind: bodyDescriptor.kind,
            blobParts: bodyDescriptor.blobParts ?? [],
            contentType: headers.get('content-type') ?? bodyDescriptor.contentType ?? '',
        }

        const requestId = bodyDescriptor.kind === 'bytes'
            ? ops.op_request_create(metadata, bodyDescriptor.bytes ?? new Uint8Array())
            : ops.op_request_create(metadata)

        const snapshot = createSnapshot(metadata, bodyDescriptor, headers)
        return createRequestFromSnapshot(requestId, snapshot)
    }

    get url() {
        return this._url
    }

    get method() {
        return this._method
    }

    get headers() {
        return this[kHeaders]
    }

    get credentials() {
        return this._credentials
    }

    get mode() {
        return this._mode
    }

    get cache() {
        return this._cache
    }

    get redirect() {
        return this._redirect
    }

    get referrer() {
        return this._referrer
    }

    get integrity() {
        return this._integrity
    }

    get keepalive() {
        return this._keepalive
    }

    get bodyUsed() {
        return this[kBodyUsed]
    }

    async arrayBuffer() {
        const bytes = await consumeBodyBytes(this)
        return bytes.buffer.slice(bytes.byteOffset, bytes.byteOffset + bytes.byteLength)
    }

    async text() {
        const buffer = await this.arrayBuffer()
        return decoder.decode(buffer)
    }

    async json() {
        const text = await this.text()
        return JSON.parse(text)
    }

    async blob() {
        const bytes = await consumeBodyBytes(this)
        return new Blob([bytes], { type: this._contentType ?? this.headers.get('content-type') ?? '' })
    }

    clone() {
        if (this.bodyUsed) {
            throw new TypeError('Cannot clone a Request with disturbed body')
        }

        const clonedId = ops.op_request_clone(this[kRequestId])
        const snapshot = ops.op_request_snapshot(clonedId)
        return createRequestFromSnapshot(clonedId, snapshot)
    }
}

export { Request }

File: lib/web/request.go
package web

import (
	"fmt"
	"sync"

	"github.com/comoland/como/js"
)

const (
	bodyKindNone  = "none"
	bodyKindBytes = "bytes"
	bodyKindBlob  = "blob"
)

type requestRecord struct {
	id          string
	url         string
	method      string
	headers     map[string]string
	credentials string
	mode        string
	cache       string
	redirect    string
	referrer    string
	integrity   string
	keepalive   bool
	bodyKind    string
	body        []byte
	blobParts   []string
	contentType string
	bodyUsed    bool
}

func (r *requestRecord) cloneHeaders() map[string]string {
	if len(r.headers) == 0 {
		return map[string]string{}
	}

	cloned := make(map[string]string, len(r.headers))
	for k, v := range r.headers {
		cloned[k] = v
	}
	return cloned
}

func (r *requestRecord) hasBody() bool {
	return r.bodyKind != bodyKindNone
}

func (r *requestRecord) cloneWithoutID() *requestRecord {
	var bodyCopy []byte
	if len(r.body) > 0 {
		bodyCopy = make([]byte, len(r.body))
		copy(bodyCopy, r.body)
	}

	var blobPartsCopy []string
	if len(r.blobParts) > 0 {
		blobPartsCopy = make([]string, len(r.blobParts))
		copy(blobPartsCopy, r.blobParts)
	}

	return &requestRecord{
		url:         r.url,
		method:      r.method,
		headers:     r.cloneHeaders(),
		credentials: r.credentials,
		mode:        r.mode,
		cache:       r.cache,
		redirect:    r.redirect,
		referrer:    r.referrer,
		integrity:   r.integrity,
		keepalive:   r.keepalive,
		bodyKind:    r.bodyKind,
		body:        bodyCopy,
		blobParts:   blobPartsCopy,
		contentType: r.contentType,
		bodyUsed:    false,
	}
}

type requestStore struct {
	mu      sync.RWMutex
	records map[string]*requestRecord
}

func newRequestStore() *requestStore {
	return &requestStore{
		records: make(map[string]*requestRecord),
	}
}

func (s *requestStore) insert(record *requestRecord) string {
	s.mu.Lock()
	defer s.mu.Unlock()

	record.id = generateUUID()
	s.records[record.id] = record
	return record.id
}

func (s *requestStore) snapshot(id string) (*requestRecord, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	record, ok := s.records[id]
	if !ok {
		return nil, fmt.Errorf("request not found: %s", id)
	}

	copyRecord := record.cloneWithoutID()
	copyRecord.id = record.id
	copyRecord.bodyUsed = record.bodyUsed
	return copyRecord, nil
}

func (s *requestStore) clone(id string) (*requestRecord, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	record, ok := s.records[id]
	if !ok {
		return nil, fmt.Errorf("request not found: %s", id)
	}

	if record.bodyUsed {
		return nil, fmt.Errorf("cannot clone a disturbed Request")
	}

	cloned := record.cloneWithoutID()
	cloned.id = generateUUID()
	s.records[cloned.id] = cloned
	return cloned, nil
}

func (s *requestStore) consumeBody(id string) ([]byte, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	record, ok := s.records[id]
	if !ok {
		return nil, fmt.Errorf("request not found: %s", id)
	}

	if record.bodyUsed {
		return nil, fmt.Errorf("body already used")
	}

	data, err := record.readBody()
	if err != nil {
		return nil, err
	}

	record.bodyUsed = true
	// Store concrete bytes for potential future snapshots/clones
	if len(data) > 0 {
		record.bodyKind = bodyKindBytes
		record.body = make([]byte, len(data))
		copy(record.body, data)
		record.blobParts = nil
	} else {
		record.body = nil
		record.blobParts = nil
		record.bodyKind = bodyKindNone
	}

	return data, nil
}

func (r *requestRecord) readBody() ([]byte, error) {
	switch r.bodyKind {
	case bodyKindNone:
		return []byte{}, nil
	case bodyKindBytes:
		if len(r.body) == 0 {
			return []byte{}, nil
		}
		data := make([]byte, len(r.body))
		copy(data, r.body)
		return data, nil
	case bodyKindBlob:
		if len(r.blobParts) == 0 {
			return []byte{}, nil
		}

		var result []byte
		for _, partID := range r.blobParts {
			part, err := globalBlobStore.GetPart(partID)
			if err != nil {
				return nil, err
			}

			chunk, err := part.Read()
			if err != nil {
				return nil, err
			}
			result = append(result, chunk...)
		}
		return result, nil
	default:
		return nil, fmt.Errorf("unknown body kind: %s", r.bodyKind)
	}
}

type requestCreatePayload struct {
	URL         string            `mapstructure:"url"`
	Method      string            `mapstructure:"method"`
	Headers     map[string]string `mapstructure:"headers"`
	Credentials string            `mapstructure:"credentials"`
	Mode        string            `mapstructure:"mode"`
	Cache       string            `mapstructure:"cache"`
	Redirect    string            `mapstructure:"redirect"`
	Referrer    string            `mapstructure:"referrer"`
	Integrity   string            `mapstructure:"integrity"`
	Keepalive   bool              `mapstructure:"keepalive"`
	BodyKind    string            `mapstructure:"bodyKind"`
	BlobParts   []string          `mapstructure:"blobParts"`
	ContentType string            `mapstructure:"contentType"`
}

var globalRequestStore = newRequestStore()

func registerRequest(ctx *js.Context) {
	mod := ctx.NewModule("request.go")
	store := globalRequestStore

	mod.Export("op_request_create", func(args js.Arguments) interface{} {
		if args.Len() == 0 {
			return ctx.Throw("op_request_create requires payload")
		}

		payload := requestCreatePayload{}
		if err := args.GetMap(0, &payload); err != nil {
			return ctx.Throw(err.Error())
		}

		if payload.URL == "" {
			return ctx.Throw("Request URL is required")
		}

		if payload.Method == "" {
			payload.Method = "GET"
		}

		if payload.Headers == nil {
			payload.Headers = map[string]string{}
		}

		bodyKind := payload.BodyKind
		if bodyKind == "" {
			bodyKind = bodyKindNone
		}

		var bodyBytes []byte
		switch bodyKind {
		case bodyKindNone:
			// no-op
		case bodyKindBytes:
			if args.Len() < 2 {
				return ctx.Throw("body bytes are required for bodyKind 'bytes'")
			}
			raw, err := args.GetTypedArray(1)
			if err != nil {
				return ctx.Throw("body must be Uint8Array")
			}
			bodyBytes = make([]byte, len(raw))
			copy(bodyBytes, raw)
		case bodyKindBlob:
			if len(payload.BlobParts) == 0 {
				return ctx.Throw("blob body requires part IDs")
			}
		default:
			return ctx.Throw(fmt.Sprintf("unsupported bodyKind '%s'", bodyKind))
		}

		record := &requestRecord{
			url:         payload.URL,
			method:      payload.Method,
			headers:     payload.Headers,
			credentials: payload.Credentials,
			mode:        payload.Mode,
			cache:       payload.Cache,
			redirect:    payload.Redirect,
			referrer:    payload.Referrer,
			integrity:   payload.Integrity,
			keepalive:   payload.Keepalive,
			bodyKind:    bodyKind,
			body:        bodyBytes,
			blobParts:   payload.BlobParts,
			contentType: payload.ContentType,
			bodyUsed:    false,
		}

		return store.insert(record)
	})

	mod.Export("op_request_snapshot", func(args js.Arguments) interface{} {
		if args.Len() == 0 {
			return ctx.Throw("op_request_snapshot requires request id")
		}

		id, ok := args.Get(0).(string)
		if !ok {
			return ctx.Throw("request id must be a string")
		}

		record, err := store.snapshot(id)
		if err != nil {
			return ctx.Throw(err.Error())
		}

		return map[string]interface{}{
			"id":          record.id,
			"url":         record.url,
			"method":      record.method,
			"headers":     record.headers,
			"credentials": record.credentials,
			"mode":        record.mode,
			"cache":       record.cache,
			"redirect":    record.redirect,
			"referrer":    record.referrer,
			"integrity":   record.integrity,
			"keepalive":   record.keepalive,
			"bodyUsed":    record.bodyUsed,
			"hasBody":     record.hasBody(),
			"contentType": record.contentType,
		}
	})

	mod.Export("op_request_clone", func(args js.Arguments) interface{} {
		if args.Len() == 0 {
			return ctx.Throw("op_request_clone requires request id")
		}

		id, ok := args.Get(0).(string)
		if !ok {
			return ctx.Throw("request id must be string")
		}

		cloned, err := store.clone(id)
		if err != nil {
			return ctx.Throw(err.Error())
		}

		return cloned.id
	})

	mod.Export("op_request_consume_body", func(args js.Arguments) interface{} {
		if args.Len() == 0 {
			return ctx.Throw("op_request_consume_body requires request id")
		}

		id, ok := args.Get(0).(string)
		if !ok {
			return ctx.Throw("request id must be string")
		}

		data, err := store.consumeBody(id)
		if err != nil {
			return ctx.Throw(err.Error())
		}

		return data
	})
}
